### PR TITLE
Rework filters for performance, technical debt, and usability. Issue: #1866

### PR DESCRIPTION
**Pull request by garethbowen**
This moves the document filters down in one level in scope from Inbox. The idea is that we shouldn't need to initialise filters for the Reports tab when viewing Contacts, and vice versa. Furthermore the filters will now be cleared when you change tabs which meets users expectations.

### CODE REVIEW

File: static/js/controllers/contacts.js [hunk 1]
```diff
@@ -151,6 +142,44 @@ var _ = require('underscore'),
...
+      $scope.resetFilterModel = function() {
+        $scope.filters = {};
+        SearchFilters.reset();
+        $scope.search();
+      };
+
+      $scope.search();
+
+      $scope.$on('export', function() {
+        if ($scope.currentTab === 'contacts') {
```
Comment (SCdF) [inline static/js/controllers/contacts.js]: _(apologies if this is in the tests or explained elsewhere)_

How could this fail? How could this function be ran but the current tab not be contacts? Does this `$on` get setup at angular init, and so it being in contacts controller is merely a useful code sorting thing for humans and everyone's `$on`s get run every time, or does angular only have that attached when the contacts controller is active?

File: templates/partials/configuration_filters.html [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    <ul role="menu" aria-labelledby="report" class="dropdown-menu mm-dropdown-menu">
+      <li class="ng-scope" role="presentation" mm-auth="can_configure" ng-repeat="page in configurationPages">
+        <a ui-sref="{{page.state}}" role="menuitem" tabindex="-1">
+          <i class="fa fa-fw {{page.icon}}"></i>
+          <span translate>{{::page.name}}</span>
+        </a>
+      </li>
+    </ul>
+  </span>
+</div>
```
Comment (SCdF) [inline templates/partials/configuration_filters.html]: :sob: image :sob:

File: static/js/services/generate-search-requests.js [hunk 3]
```diff
@@ -7,7 +7,7 @@ var _ = require('underscore'),
 
   var inboxServices = angular.module('inboxServices');
 
-  var endOfAlphabet = '\ufff0';
+  var END_OF_ALPHABET = '\ufff0';
```
Comment (SCdF) [inline static/js/services/generate-search-requests.js]: :+1:

File: static/js/controllers/contacts.js (hunk 1 above)
Comment (garethbowen) [inline static/js/controllers/contacts.js]: That's me being defensive. I'm pretty sure it'll get unloaded so won't be invoked, but there could be a race condition.

File: static/js/services/search-filters.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  inboxServices.factory('SearchFilters', ['$translate',
+    function($translate) {
+      
+      var initFreetext = function(callback) {
+        $('#search').on('click', function(e) {
+          e.preventDefault();
+          callback();
+        });
+        $('#freetext').on('keypress', function(e) {
+          if (e.which === 13) {
```
Comment (SCdF) [inline static/js/services/search-filters.js]: Magic number

Comment (alxndrsn) [inline static/js/services/search-filters.js]: I think that one's unlucky, rather than magic.

**Pull request merged**